include/boost/redis/detail/exec_fsm.hpp

```diff
@@ -0,0 +1,148 @@
+
+#include <boost/asio/cancellation_type.hpp>
+#include <boost/asio/error.hpp>
+#include <boost/assert.hpp>
+#include <boost/system/error_code.hpp>
+
+#include <cstddef>
+#include <memory>
+
+// Sans-io algorithm for async_exec, as a finite state machine
+
+namespace boost::redis::detail {
+
+// What should we do next?
+enum class exec_action_type
+{
+   immediate,          // Invoke asio::async_immediate to avoid re-entrancy problems
+   done,               // Call the final handler
+   write,              // Notify the writer task
+   wait_for_response,  // Wait to be notified
+   cancel_run,         // Cancel the connection's run operation
+};
+
+class exec_action {
+   exec_action_type type_;
+   system::error_code ec_;
+   std::size_t bytes_read_;
+
+public:
+   exec_action(exec_action_type type) noexcept
+   : type_{type}
+   { }
+
+   exec_action(system::error_code ec, std::size_t bytes_read = 0u) noexcept
+   : type_{exec_action_type::done}
+   , ec_{ec}
+   , bytes_read_{bytes_read}
+   { }
+
+   exec_action_type type() const { return type_; }
+   system::error_code error() const { return ec_; }
+   std::size_t bytes_read() const { return bytes_read_; }
+};
+
+class exec_fsm {
+   int resume_point_{0};
+   multiplexer* mpx_{nullptr};
+   std::shared_ptr<multiplexer::elem> elem_;
+
+   static bool is_cancellation(asio::cancellation_type_t type)
+   {
+      return !!(
+         type & (asio::cancellation_type_t::total | asio::cancellation_type_t::partial |
+                 asio::cancellation_type_t::terminal));
+   }
+
+   exec_action resume_impl(bool connection_is_open, asio::cancellation_type_t cancel_state)
+   {
+      switch (resume_point_) {
+         BOOST_REDIS_CORO_INITIAL
+
+         // Check whether the user wants to wait for the connection to
+         // be established.
+         if (elem_->get_request().get_config().cancel_if_not_connected && !connection_is_open) {
+            BOOST_REDIS_YIELD(resume_point_, 1, exec_action_type::immediate)
+            return system::error_code(error::not_connected);
+         }
+
+         // Add the request to the multiplexer
+         mpx_->add(elem_);
+
+         // Notify the writer task that there is work to do. If the task is not
+         // listening (e.g. it's already writing or the connection is not healthy),
+         // this is a no-op. Since this is sync, no cancellation can happen here.
+         BOOST_REDIS_YIELD(resume_point_, 2, exec_action_type::write)
+
+         while (true) {
+            // Wait until we get notified. This will return once the request completes,
+            // or upon any kind of cancellation
+            BOOST_REDIS_YIELD(resume_point_, 3, exec_action_type::wait_for_response)
+
+            // If the request has completed (with error or not), we're done
+            if (elem_->is_done()) {
+               return exec_action{
+                  elem_->get_error(),
+                  elem_->get_error() ? 0u : elem_->get_read_size()};
```

Comment:
I think here you could return whatever is in `elem_->get_read_size()`. Seting to `0` when there is an error seems unnecessary and might indeed erase useful information.